src/migrations/20240611185509-goalStatusChanges-historical.js

```diff
@@ -0,0 +1,191 @@
+				('Suspended', 'In Progress', 'reviewReport', 'UPDATE', 'Activity Report reviewed', NULL),
+				('Suspended', 'Draft', 'createGoalsForReport', 'UPDATE', NULL, NULL),
+				('Closed', 'Not Started', NULL, 'UPDATE', NULL, NULL),
+				('Not Started', 'Draft', 'createGoals', 'UPDATE', 'Goal creation', NULL),
+				('In Progress', 'Closed', 'dup_goals_Update_Goals', 'UPDATE', 'Migration', NULL),
+				(NULL, 'Draft', NULL, 'UPDATE', NULL, NULL),
+				('Not Started', 'In Progress', 'mergeGoalHandler', 'UPDATE', NULL, NULL),
+				('In Progress', 'Suspended', 'dup_goals_Update_Goals', 'UPDATE', 'Migration', NULL),
+				('In Progress', 'Not Started', 'RUN MIGRATIONS', 'UPDATE', 'Migration', NULL),
+				('Draft', NULL, 'deleteGoal', 'DELETE', NULL, NULL),
+				('In Progress', 'Draft', NULL, 'UPDATE', NULL, NULL),
+				('Closed', NULL, 'createGoalsForReport', 'DELETE', NULL, NULL),
+				('In Progress', 'Draft', 'createGoals', 'UPDATE', 'Goal creation', NULL),
+				(NULL, 'In Progress', NULL, 'UPDATE', NULL, NULL),
+				('Suspended', 'In Progress', 'createGoals', 'UPDATE', 'Goal creation', NULL),
+				('Draft', 'In Progress', 'createGoals', 'UPDATE', 'Goal creation', NULL),
+				('Not Started', 'Draft', NULL, 'UPDATE', NULL, NULL),
+				('Suspended', 'In Progress', 'createGoalsForReport', 'UPDATE', NULL, NULL)
+			) AS tmp_data("oldStatus", "newStatus", descriptor, dml_type, reason, context)
+	),
+    status_changes_query AS (
+        SELECT
+            zg.data_id AS "goalId",
+            zg.dml_as AS "userId",
+            u.name AS "userName",
+            ARRAY_AGG(DISTINCT r.name) FILTER (WHERE r.name IS NOT NULL) AS "userRoles",
+            zg.old_row_data ->> 'status' AS "oldStatus",
+            zg.new_row_data ->> 'status' AS "newStatus",
+            COALESCE((ARRAY_AGG(TRIM(NULLIF(zg.new_row_data ->> 'closeSuspendReason', ''))))[1], rl.reason) AS "reason",
+            COALESCE((ARRAY_AGG(TRIM(NULLIF(zg.new_row_data ->> 'closeSuspendContext', ''))))[1], rl.context) AS "context",
+            zg.dml_timestamp AS "createdAt",
+            zg.dml_timestamp AS "updatedAt"
+        FROM
+            "ZALGoals" zg
+        LEFT JOIN "ZADescriptor" zd ON zg.descriptor_id = zd.id
+        LEFT JOIN "Users_timeseries" u ON zg."dml_as" = u.data_id AND zg.dml_timestamp BETWEEN u.timeband_start AND u.timeband_end
+        LEFT JOIN "UserRoles_timeseries" ur ON zg."dml_as" = ur."userId" AND zg.dml_timestamp BETWEEN ur.timeband_start AND ur.timeband_end
+        LEFT JOIN "Roles" r ON ur."roleId" = r.id
+        LEFT JOIN reason_lookup rl
+            ON (zg.old_row_data ->> 'status' = rl."oldStatus")
+            AND (zg.new_row_data ->> 'status' = rl."newStatus")
+            AND (zd.descriptor = rl.descriptor)
+            AND (zg.dml_type::text = rl.dml_type)
+        WHERE
+            (NULLIF(TRIM(zg.old_row_data ->> 'status'), '') IS NOT NULL
+                OR NULLIF(TRIM(zg.new_row_data ->> 'status'), '') IS NOT NULL)
```

Comment:
I think you need some more logic here to avoid re-inserting status change records that already exist.

Reply:
yea..... your right

Reply:
this could be a problem as I don't really know how to do that as the times might not line up exactly